airbyte-integrations/connectors/source-mysql/src/main/kotlin/io/airbyte/integrations/source/mysql/MysqlJdbcPartitionFactory.kt

```diff
@@ -158,11 +158,22 @@ class MysqlJdbcPartitionFactory(
     override fun create(streamFeedBootstrap: StreamFeedBootstrap): MysqlJdbcPartition? {
         val stream: Stream = streamFeedBootstrap.feed
         val streamState: DefaultJdbcStreamState = streamState(streamFeedBootstrap)
-        val opaqueStateValue: OpaqueStateValue =
-            when (streamFeedBootstrap.currentState?.isEmpty) {
-                false -> streamFeedBootstrap.currentState!!
-                else -> return coldStart(streamState)
-            }
+
+        // An empty table stream state will be marked as a nullNode. This prevents repeated attempt
+        // to read it
+        if (streamFeedBootstrap.currentState?.isNull == true) {
+            return null
+        }
+
+        // A legacy saved state may be null for an empty table. We will attempt to read it again
+        if (
+            streamFeedBootstrap.currentState == null ||
```

Comment:
is there any difference between this check and the check above

```
streamFeedBootstrap.currentState?.isNull == true
```

Reply:
empty is the case a CDC state if no longer on the server + configured to re-sync rather than fail: we mark this state as `{}` which is empty, but not null.

I'm checking if we need to check also for `.isNull`

Reply:
this flow is necessary for both backward compatibility: upgrading from legacy mysql that saved a state for an empty table as a null node, and also for all flows of new mysql.
Node line 164 - it provides an exit condition to prevent infinite attempts to read an empty table, but also ensures that on every sync we try to read an empty table once in case there are newly inserted rows.

Reply:
In full refresh a new sync starts with no state for the full refresh table:
streamFeedBootstrap.currentState == null ---→ cold start (read from top)

and after an empty table is read the state will become a NullNode:
streamFeedBootstrap.currentState?.isNull == true ----→ return null (finish sync)